File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/location/replay/ParseGpxTask.java
package com.mapbox.services.android.navigation.v5.location.replay;

import android.location.Location;
import android.os.AsyncTask;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.xml.sax.SAXException;

import java.io.IOException;
import java.io.InputStream;
import java.text.ParseException;
import java.util.List;

import javax.xml.parsers.ParserConfigurationException;

class ParseGpxTask extends AsyncTask<InputStream, Void, List<Location>> {

  private static final int FIRST_INPUT_STREAM = 0;

  private Listener listener;
  private GpxParser parser;

  ParseGpxTask(GpxParser parser, Listener listener) {
    this.parser = parser;
    this.listener = listener;
  }

  @Override
  protected List<Location> doInBackground(InputStream... inputStreams) {
    InputStream inputStream = inputStreams[FIRST_INPUT_STREAM];
    try {
      return parseGpxStream(inputStream);
    } catch (IOException exception) {
      listener.onParseError(exception);
      return null;
    }
  }

  @Override
  protected void onPostExecute(List<Location> locationList) {
    if (locationList != null && !locationList.isEmpty()) {
      listener.onParseComplete(locationList);
    } else {
      listener.onParseError(new RuntimeException("An error occurred parsing the GPX Xml."));
    }
  }

  @Nullable
  private List<Location> parseGpxStream(InputStream inputStream) throws IOException {
    try {
      return parser.parseGpx(inputStream);
    } catch (ParserConfigurationException | ParseException | SAXException | IOException exception) {
      exception.printStackTrace();
      listener.onParseError(exception);
      return null;
    } finally {
      inputStream.close();
    }
  }

  public interface Listener {

    void onParseComplete(@NonNull List<Location> gpxLocationList);

    void onParseError(Exception exception);
  }
}

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/location/replay/ParseGpxTask.kt
package com.mapbox.services.android.navigation.v5.location.replay

import android.location.Location
import android.os.AsyncTask

import org.xml.sax.SAXException

import java.io.IOException
import java.io.InputStream
import java.text.ParseException

import javax.xml.parsers.ParserConfigurationException

internal class ParseGpxTask(private val parser: GpxParser, private val listener: Listener) : AsyncTask<InputStream, Void, List<Location>>() {

    override fun doInBackground(vararg inputStreams: InputStream): List<Location>? {
        val inputStream = inputStreams[FIRST_INPUT_STREAM]
        try {
            return parseGpxStream(inputStream)
        } catch (exception: IOException) {
            listener.onParseError(exception)
            return null
        }

    }

    override fun onPostExecute(locationList: List<Location>?) {
        if (locationList != null && !locationList.isEmpty()) {
            listener.onParseComplete(locationList)
        } else {
            listener.onParseError(RuntimeException("An error occurred parsing the GPX Xml."))
        }
    }

    @Throws(IOException::class)
    private fun parseGpxStream(inputStream: InputStream): List<Location>? {
        try {
            return parser.parseGpx(inputStream)
        } catch (exception: ParserConfigurationException) {
            exception.printStackTrace()
            listener.onParseError(exception)
            return null
        } catch (exception: ParseException) {
            exception.printStackTrace()
            listener.onParseError(exception)
            return null
        } catch (exception: SAXException) {
            exception.printStackTrace()
            listener.onParseError(exception)
            return null
        } catch (exception: IOException) {
            exception.printStackTrace()
            listener.onParseError(exception)
            return null
        } finally {
            inputStream.close()
        }
    }

    interface Listener {

        fun onParseComplete(gpxLocationList: List<Location>)

        fun onParseError(exception: Exception)
    }

    companion object {

        private val FIRST_INPUT_STREAM = 0
    }
}
